pkg/api/handlers/util.go

```diff
@@ -109,3 +112,42 @@ func nodePoolEqualsWithScaling(old, new models.NodePool) error {
 
 	return nil
 }
+
+func fetchOpenstackMetadata(request *http.Request, principal *models.Principal) (*models.OpenstackMetadata, error) {
+	tokenID := request.Header.Get("X-Auth-Token")
+
+	authOptions := &tokens.AuthOptions{
+		IdentityEndpoint: principal.AuthURL,
+		TokenID:          tokenID,
+		AllowReauth:      true,
+		Scope: tokens.Scope{
+			ProjectID: principal.Account,
+		},
+	}
+
+	client, err := openstack.NewSharedOpenstackClientFactory(nil, nil, nil, getTracingLogger(request)).ProjectClientFor(authOptions)
```

Comment:
This will break. We [cache the client](https://github.com/sapcc/kubernikus/blob/master/pkg/client/openstack/factory.go#L125) based on the project id (which is stupid in the first place because there might be more than one cluster eventually). But this means the apiserver will keep the client around reusing the first token and ignore the token that was passed in the current reques, once the token is expired this call will fail until the process is restarted.

Reply:
The factory is recreated for each request. Ugly but the cache is only short-lived. 🤷‍♂️

Reply:
I see. Ok so we can leave it like this for now. 👍